### DISCUSSION TITLE
how to set timeout when execute new process?

### DISCUSSION THREAD

**Discussion by itech001** [google/zx#259] [Help needed] [answered]
how to set timeout when execute new process?

**Comment by mblarsen** [comment]
`$` returns a [`ProcessPromise`](https://github.com/google/zx#processpromise). On it is a `kill()` function that you can use

E.g.

```js
#!/usr/bin/env zx

const timeout = (p, ms) => {
  return new Promise((resolve, reject) => {
    p.then(resolve).catch(reject);
    setTimeout(() => {
      p.kill()
      reject(new Error('timeout'));
    }, ms);
  })
}

const output = await timeout($`sleep 5; ls`, 2000)

console.log(output)
```

output: 

```shell
§ zx timeout.mjs
$ sleep 5; ls
file:///.../timeout.mjs:8
      reject(new Error('timeout'));
             ^

Error: timeout
    at Timeout._onTimeout (file:///.../timeout.mjs:8:14)
    at listOnTimeout (node:internal/timers:557:17)
    at processTimers (node:internal/timers:500:7)
```

Changing the command to `sleep 2; false` with timeout 5000 you'll see that the normal reject state also works:

```shell
$ sleep 2; false
Error:
    at file:///Users/mbl/timeout.mjs:13:31
    exit code: 1
```

**Reply by antongolub** [reply]
In v[6.0.2](https://github.com/google/zx/releases/tag/6.0.2)...[6.0.3](https://github.com/google/zx/releases/tag/6.0.3) we've introduced a shiny new experimental API for this:
```js
import {withTimeout} from 'zx/experimental'
await withTimeout(100, 'SIGTERM')`sleep 9999`
```
And there's also `--experimantal` flag to enable proposals via CLI: `zx --experimental script.mjs`

https://github.com/google/zx#withtimeout

**Reply by mblarsen** [reply]
Nice, even better

**Comment by andyfleming** [comment, marked as answer]
The latest syntax is...

```ts
await $`sleep 999`.timeout('5s')

// Or with a specific signal.
await $`sleep 999`.timeout('5s', 'SIGKILL')
```

Updated docs are here: https://google.github.io/zx/process-promise#timeout
